File: src/services/fruit-picker-client.js
const axios = require('axios');
const config = require('../core/config');
const logger = require('../utils/logger');

class FruitPickerClient {
  constructor() {
    this.baseUrl = config.services.fruitPicker.url;
    this.apiKey = config.services.fruitPicker.apiKey;
    this.timeout = 10000;
  }

  async pickFruit(fruitType, quantity) {
    try {
      const response = await axios.post(
        `${this.baseUrl}/api/v1/pick-fruit`,
        {
          fruitType: fruitType,
          quantity: quantity,
          quality: 'premium'
        },
        {
          headers: {
            'X-API-Key': this.apiKey
          },
          timeout: this.timeout
        }
      );
      
      logger.info(`Fruit picker job created: ${response.data.jobId}`);
      return response.data;
    } catch (error) {
      if (error.code === 'ECONNABORTED') {
        logger.error('Fruit picker service timeout');
        throw new Error('Fruit picker service timeout');
      }
      logger.error(`Failed to pick fruit: ${error.message}`);
      throw error;
    }
  }

  async getJobStatus(jobId) {
    try {
      const response = await axios.get(
        `${this.baseUrl}/api/v1/jobs/${jobId}`,
        {
          headers: {
            'X-API-Key': this.apiKey
          },
          timeout: 5000
        }
      );
      return response.data;
    } catch (error) {
      logger.error(`Failed to get job status: ${error.message}`);
      throw error;
    }
  }
}

module.exports = FruitPickerClient;

File: src/services/fruit_picker_client.py
import time
import structlog
import httpx
from src.core.config import settings

logger = structlog.get_logger()


class FruitPickerClient:
    def __init__(self):
        self.base_url = settings.fruit_picker_url
        self.api_key = settings.fruit_picker_api_key

    def pick_fruit(self, fruit_type: str, quantity: int, quality: str = "premium") -> dict:
        attempt = 0
        last_error = None

        while attempt < settings.max_retries:
            try:
                response = httpx.post(
                    f"{self.base_url}/api/v1/pick-fruit",
                    json={"fruitType": fruit_type, "quantity": quantity, "quality": quality},
                    headers={"X-API-Key": self.api_key},
                    timeout=10,
                )
                response.raise_for_status()
                return response.json()

            except httpx.TimeoutException as e:
                logger.error("fruit_picker_timeout", attempt=attempt, error=str(e))
                last_error = e
                attempt += 1
                # TODO: Implement circuit breaker pattern here
                time.sleep(settings.retry_delay)

            except httpx.HTTPStatusError as e:
                logger.error("fruit_picker_error", status=e.response.status_code, error=str(e))
                raise

        raise Exception(f"Fruit picker failed after {settings.max_retries} attempts: {last_error}")

    def get_job_status(self, job_id: str) -> dict:
        response = httpx.get(
            f"{self.base_url}/api/v1/jobs/{job_id}",
            headers={"X-API-Key": self.api_key},
            timeout=10,
        )
        response.raise_for_status()
        return response.json()
